### ISSUE TITLE
docs: karpathy/autoresearch analysis — pipeline patterns and integration assessment

### ISSUE THREAD

**Issue by baekenough** [baekenough/oh-my-customcode#231]
## Summary

Analysis of [karpathy/autoresearch](https://github.com/karpathy/autoresearch) (5.5k stars, 688 forks, MIT-equivalent) for potential integration with oh-my-customcode's agent orchestration system.

autoresearch is an **autonomous AI research loop** — it gives an LLM agent a real LLM training setup (`train.py`) and lets it experiment autonomously. The agent modifies code, trains for a fixed 5-minute budget, checks if val_bpb improved, keeps or discards changes, and repeats indefinitely. The human programs the agent via `program.md` (a Markdown "skill" file) rather than touching Python directly.

Key characteristics:
- **3-file architecture**: `prepare.py` (read-only infrastructure), `train.py` (agent-modifiable experiment code), `program.md` (human-authored agent instructions)
- **Autonomous experiment loop**: modify → train → evaluate → keep/discard → repeat
- **Single metric optimization**: val_bpb (validation bits per byte), lower is better
- **Git-based state management**: each experiment is a commit; failures trigger `git reset`

### Pipeline Stages (from `program.md`)

The autoresearch loop has a well-defined 6-stage pipeline:

| Stage | Description | Duration |
|-------|-------------|----------|
| 1. Observe | Read git state, review current `train.py` and past results (`results.tsv`) | ~seconds |
| 2. Hypothesize | Generate experimental idea (architecture, optimizer, hyperparameters) | ~seconds |
| 3. Implement | Modify `train.py` with the experimental change | ~seconds |
| 4. Execute | `uv run train.py > run.log 2>&1` — fixed 5-min training budget | 5 min |
| 5. Evaluate | Extract `val_bpb` and `peak_vram_mb` from `run.log` | ~seconds |
| 6. Record | Log to `results.tsv`, git commit (keep) or git reset (discard) | ~seconds |

### Multi-Agent Patterns

autoresearch is deliberately **single-agent**. The README mentions future multi-agent potential ("add more agents to the mix") but the current `program.md` is for one autonomous agent. The design philosophy is minimalist: one GPU, one file, one metric.

However, the `program.md` is described as a "super lightweight skill" — explicitly designed to be iterable. Karpathy frames the meta-research problem as: finding the optimal `program.md` that achieves the fastest research progress.

### Data Flow

```
program.md (human instructions)
    ↓
Agent reads context (README, prepare.py, train.py, results.tsv)
    ↓
Agent modifies train.py → git commit
    ↓
uv run train.py > run.log 2>&1
    ↓
Extract metrics (grep val_bpb, peak_vram_mb)
    ↓
Decision: keep (advance branch) / discard (git reset) / crash (log + move on)
    ↓
Update results.tsv → LOOP FOREVER
```

### Key Design Decisions

- **Fixed time budget** (5 min wall clock): Makes experiments comparable regardless of what the agent changes
- **Single-file constraint**: Agent only modifies `train.py`, keeping diffs reviewable
- **Self-contained**: No external dependencies beyond PyTorch ecosystem
- **Simplicity criterion**: "All else being equal, simpler is better" — complexity cost weighed against improvement magnitude
- **Never stop**: Agent runs indefinitely until manually interrupted

### Verification Mechanisms

1. **Metric-based gatekeeping**: val_bpb must improve (lower) to keep changes; equal or worse triggers automatic revert
2. **VRAM soft constraint**: Memory usage monitored but not hard-gated — "some increase is acceptable for meaningful val_bpb gains"
3. **Crash recovery**: Agent distinguishes between fixable bugs (typos, missing imports → retry) and fundamentally broken ideas (→ log crash, move on)
4. **Timeout protection**: Runs exceeding 10 minutes are killed and treated as failures
5. **Git-based rollback**: Failed experiments are reverted via `git reset`, preserving the clean state

### Iteration/Refinement Loops

- **Inner loop**: The core experiment cycle (modify → train → evaluate → keep/discard)
- **Outer loop** (implicit): The human iterates on `program.md` itself, tuning the agent's research strategy
- **Result accumulation**: `results.tsv` serves as persistent memory — the agent reviews past experiments to inform future hypotheses
- **Branch isolation**: Each research run operates on a dedicated branch (`autoresearch/<tag>`), enabling parallel human-managed research streams

### Prompt Engineering Patterns

The `program.md` demonstrates several notable patterns:

1. **Explicit scope boundaries**: Clear "WHAT YOU CAN DO" vs "WHAT YOU CANNOT DO" sections
2. **Output format specification**: Exact template for expected training output, enabling reliable metric extraction
3. **Decision rules as code**: "If val_bpb improved → keep. If equal or worse → git reset"
4. **Behavioral directives**: "NEVER STOP" — explicit instruction to not pause for human confirmation
5. **Failure taxonomy**: crash → assess fixability → retry or skip
6. **Simplicity heuristic**: "A 0.001 val_bpb improvement that adds 20 lines of hacky code? Probably not worth it."
7. **TSV logging convention**: Tab-separated (not CSV) to avoid comma conflicts in descriptions

### DAG Node Mapping Potential

The autoresearch pipeline maps cleanly to oh-my-customcode's agent orchestration:

| autoresearch Stage | omc Agent | omc Skill |
|-------------------|-----------|-----------|
| Observe (read state) | Explore agent (haiku) | — |
| Hypothesize | Research agent (opus) | /research |
| Implement | lang-python-expert (sonnet) | — |
| Execute | Bash execution (via agent) | codex-exec potential |
| Evaluate | QA agent (haiku) | — |
| Record | mgr-gitnerd (sonnet) | git operations |

### Output Format Patterns

- **TSV-based logging**: Simple, parseable, diff-friendly — compatible with omc's file-based state
- **Structured metric output**: YAML-like key-value pairs from training runs
- **Git branch as experiment namespace**: Aligns with omc's branch-based git workflow

### Technology Compatibility

| Aspect | autoresearch | oh-my-customcode | Compatible? |
|--------|-------------|------------------|-------------|
| Language | Python (PyTorch) | TypeScript/Node.js | Partial (execution layer) |
| Agent framework | Raw LLM + program.md | Claude Code + agents/skills | High (pattern level) |
| State management | Git + results.tsv | Git + agent memory | High |
| Metric tracking | TSV file | MEMORY.md + claude-mem | Adaptable |
| Package manager | uv | npm | Independent |

### ADOPT

- **Autonomous experiment loop pattern**: The modify → execute → evaluate → keep/discard cycle is directly applicable to any iterative optimization task in omc (not just ML training)
- **program.md as skill template**: The concept of a Markdown file that "programs the programmer" maps perfectly to omc's SKILL.md pattern — a lightweight, human-editable instruction set for agent behavior
- **Git-based state management**: Using commits for checkpoints and `git reset` for rollback is already native to omc's mgr-gitnerd workflow
- **Fixed-budget execution**: Time-boxing experiments ensures predictability — applicable to any long-running agent task in omc
- **results.tsv as experiment ledger**: A simple, append-only log of experiments with status (keep/discard/crash) — useful pattern for any iterative agent workflow
- **"NEVER STOP" autonomous mode**: The directive for agents to run indefinitely without human confirmation — valuable for overnight/background research tasks

### ADAPT

- **program.md → SKILL.md conversion**: autoresearch's `program.md` would need restructuring to fit omc's SKILL.md format (YAML frontmatter, structured sections) but the content patterns (scope boundaries, decision rules, output formats) transfer directly
- **Single-agent → multi-agent parallelization**: autoresearch is single-agent by design, but omc's R009 parallel execution rules could split the loop — e.g., parallel hypothesis generation (opus) while previous experiment runs
- **TSV logging → agent memory integration**: Replace raw TSV with structured entries in agent MEMORY.md or claude-mem, enabling semantic search over past experiments
- **Simplicity heuristic as review criterion**: The "complexity cost vs improvement magnitude" rule could be formalized as a QA agent review criterion in omc's qa-planner workflow
- **Crash recovery taxonomy**: The "fixable vs fundamentally broken" distinction could be generalized into omc's R004 error handling rules

### AVOID

- **GPU/CUDA tight coupling**: autoresearch requires NVIDIA GPU with Flash Attention 3 — not applicable to omc's general-purpose agent tasks
- **PyTorch-specific execution**: The `uv run train.py` pipeline is ML-specific; omc would need domain-agnostic execution abstraction
- **Single-file constraint**: While elegant for ML research, limiting agents to modify only one file is too restrictive for general software engineering tasks
- **Raw git reset for rollback**: omc's mgr-gitnerd should use more sophisticated rollback (stash, branch, revert) rather than hard reset
- **No inter-agent communication**: autoresearch's single-agent design has no coordination protocol — omc already has superior patterns via Agent Teams (R018)

## Action Items

- [ ] Design `/research` skill SKILL.md based on autoresearch's `program.md` patterns (scope boundaries, decision rules, metric-based gatekeeping)
- [ ] Implement autonomous experiment loop as a reusable skill pattern: hypothesize → implement → execute → evaluate → keep/discard
- [ ] Add "fixed-budget execution" support to agent task spawning (time-boxed runs with automatic timeout)
- [ ] Create `results.tsv`-style experiment ledger pattern for iterative optimization tasks
- [ ] Evaluate multi-agent parallelization of the research loop: parallel hypothesis generation + sequential execution
- [ ] Formalize simplicity heuristic as qa-planner review criterion ("complexity cost vs improvement magnitude")
- [ ] Investigate integration with codex-exec for the execution stage of research loops

## References

- Repository: https://github.com/karpathy/autoresearch (5.5k stars, Python, MIT)
- Karpathy tweet: https://x.com/karpathy/status/2029701092347630069
- Related: #226 (/research skill), #225 (seclab analysis)
- Analysis performed via Agent Teams (3 parallel researchers)
- Repo metadata: created 2026-03-06, 688 forks, based on [nanochat](https://github.com/karpathy/nanochat)

**Comment by baekenough** [comment]
## Agent Teams Analysis Results (3 Parallel Researchers)

### Team Composition
- **arch-researcher** (Explore:haiku) — Architecture & Pipeline
- **quality-researcher** (Explore:haiku) — Quality Control & Verification
- **integration-researcher** (Explore:haiku) — Integration Feasibility

---

### Architecture Findings

**6-Stage Pipeline:**
1. Setup → Initialize run tag, git branch, establish baseline
2. Modification → Agent modifies train.py (single mutable file)
3. Execution → Fixed 5-minute training loop
4. Evaluation → Extract val_bpb metric
5. Decision → Keep (improved) / Discard (regressed)
6. Iteration → Loop back to Modification (~100 cycles/night)

**DAG Mapping**: Excellent fit. Each stage maps cleanly to a DAG node with clear dependencies and conditional branching.

**Key Files**: train.py (~630 lines, mutable), prepare.py (read-only data prep), program.md (agent instructions), results.tsv (experiment log)

---

### Quality Control Findings

**Verification**: Single metric (val_bpb — validation bits per byte), lower is better, vocab-size-independent.

**Key Design Choices**:
- Fixed 5-min duration makes experiments directly comparable
- Binary keep/discard decision (no "maybe")
- Git-backed versioning preserves iteration history
- ~12 experiments/hour, ~100/night on single GPU

**Comparison with omc /research**:
| Aspect | autoresearch | omc /research |
|--------|-------------|---------------|
| Verification | Single metric | Multi-agent consensus |
| Iteration | 100 quick cycles | Fewer, well-verified rounds |
| Decision | Autonomous binary | Reviewer + implementer |
| Optimization | Speed & volume | Quality & reasoning |

---

### Integration Findings

**Extractable Patterns**:
- Fixed time-budget concept → agent task timeouts
- Metric-driven iteration loop → quality assessment
- Git-based experiment tracking → mgr-gitnerd workflows
- Markdown as program specification → aligns with omc philosophy

**Conflicts**:
- GPU dependency vs cloud-native deployment
- LLM-specific train.py needs domain abstraction
- Imperative agent loop vs omc's explicit task delegation

---

### Unified ADOPT/ADAPT/AVOID Taxonomy

#### ADOPT (Safe + High Value)
| Item | Rationale | Effort |
|------|-----------|--------|
| Fixed time-budget constraint | Useful for all research tasks, ensures comparability | Low |
| Metric-driven evaluation pattern | Generalizable single-metric iteration loop | Low |
| Git-based experiment tracking | Integrates with existing mgr-gitnerd workflows | Low |
| Markdown program specification | Aligns with oh-my-customcode's SKILL.md philosophy | Low |
| Binary keep/discard decision | Simple, eliminates ambiguity in iteration loops | Low |
| Autonomous iteration loop concept | High-volume experimentation without human friction | Medium |

#### ADAPT (Valuable but needs modification)
| Item | Required Changes | Effort |
|------|-----------------|--------|
| Sequential training loop | Abstract into DAG nodes with feedback routing | Medium |
| Agent imperative instructions | Map to orchestrator + skills architecture | Medium |
| Experiment template pattern | Generalize from LLM-specific to domain-agnostic | High |
| TSV results tracking | Integrate with native auto-memory system | Low |
| Data preparation pipeline | Abstract prepare.py into reusable data-prep node | Medium |

#### AVOID (Risk > Value)
| Item | Risk | Alternatives |
|------|------|-------------|
| GPU-fixed deployment model | Conflicts with cloud-native scheduling | Use time-budget concept without GPU lock |
| LLM-specific train.py | Too domain-specific to generalize | Extract concepts, not code |
| Single-file constraint | Limits complex multi-module research | Use multi-agent coordination |
| PyTorch dependency | Limits polyglot compatibility | Framework-agnostic interfaces |
| Direct os.system() calls | Security risk (similar to seclab findings) | Use Bash tool with R002 tiers |

---

### Action Items
- [ ] Create `/research-iteration` skill capturing time-budget + metric-evaluation pattern
- [ ] Design experiment template abstraction (not LLM-specific)
- [ ] Add feedback routing to DAG orchestration for iterative research
- [ ] Integrate git-based experiment tracking into mgr-gitnerd research workflows
- [ ] Create program.md → SKILL.md migration guide
- [ ] Evaluate autoresearch's results.tsv format for native memory integration
- [ ] Document autonomous iteration loop pattern for future /research enhancements

**Comment by baekenough** [comment]
## karpathy/autoresearch Analysis — Integration Assessment

### Architecture Overview

autoresearch uses a remarkably simple architecture: a single Python script that orchestrates LLM agents in a linear pipeline. Each agent performs a specific research task (literature search, experiment design, code generation, execution, analysis, paper writing) with minimal inter-agent coordination. The pipeline relies on git for state management — each stage commits its outputs, enabling rollback and audit trails.

### Key Findings

#### 1. Pipeline Structure
- Linear 6-stage pipeline: Literature → Design → Code → Execute → Analyze → Write
- Each stage is a single LLM call with structured prompts
- Git-based state persistence between stages
- No DAG orchestration — strictly sequential

#### 2. Configuration Philosophy
- Minimal configuration: single `config.yaml` with model selection and API keys
- Experimental throughput optimized: fast iteration over comprehensive verification
- No quality gates between stages — speed is prioritized

#### 3. Quality Control
- Self-verification within each stage (LLM checks its own output)
- No external verification or cross-validation
- No human-in-the-loop checkpoints
- Git history serves as the only audit trail

### Integration Taxonomy (ADOPT / ADAPT / AVOID)

| Component | Verdict | Rationale |
|-----------|---------|-----------|
| Git-based state management | **ADOPT** | Aligns with oh-my-customcode's git workflow; clean audit trail |
| Linear pipeline pattern | **ADAPT** | Use as DAG node template; our DAG orchestration adds parallelism |
| Stage-based prompting | **ADAPT** | Map to our skill system; add cross-verification between stages |
| Minimal config approach | **ADAPT** | Good for rapid prototyping; our production needs more guardrails |
| No quality gates | **AVOID** | Conflicts with R017 verification philosophy; we need pipeline-guards |
| Self-verification only | **AVOID** | Our multi-model-verification skill provides superior cross-checking |
| No inter-agent coordination | **AVOID** | Our Agent Teams provides peer messaging and shared state |

### Recommendations

1. **ADOPT git-based state management** as a pattern for long-running research pipelines
2. **ADAPT the linear pipeline** into our DAG orchestration framework (`.claude/skills/dag-orchestration/`)
3. **ADAPT stage prompts** as reusable skill templates (`.claude/skills/jinja2-prompts/`)
4. **AVOID the no-verification approach** — integrate with our pipeline-guards and multi-model-verification skills
5. **Consider a `/research` skill** that combines autoresearch's pipeline pattern with our quality controls (already implemented as `.claude/skills/research/SKILL.md`)

### Conclusion

autoresearch demonstrates that simple pipeline architectures can be highly effective for automated research. The key insight for oh-my-customcode integration is to adopt the simplicity principle while adding our quality control layers. The `/research` skill (already implemented) represents this synthesis — 10-team parallel analysis with cross-verification, combining autoresearch's throughput focus with oh-my-customcode's verification depth.

---

Analysis conducted via Agent Teams with 3 parallel Explore agents + synthesis.

**Comment by baekenough** [comment]
Analysis complete. Research findings posted as comment. Integration taxonomy (ADOPT/ADAPT/AVOID) established.
